util/qc_stmt.go

```diff
@@ -169,85 +254,147 @@ func ParseQcStatem(extVal []byte, sought asn1.ObjectIdentifier) EtsiQcStmtIf {
-			return etsiObj
+			return handleIdEtsiQcsQcSSCD(statem, raw)
 		} else if statem.Oid.Equal(IdEtsiQcsQcEuPDS) {
-			etsiObj := EtsiQcPds{etsiBase: etsiBase{isPresent: true}}
-			rest, err := asn1.Unmarshal(statem.Any.FullBytes, &etsiObj.PdsLocations)
-			if len(rest) != 0 || err != nil {
-				etsiObj.errorInfo = "error parsing the statementInfo field"
-			} else {
-				AppendToStringSemicolonDelim(&etsiObj.errorInfo,
-					checkAsn1Reencoding(reflect.ValueOf(etsiObj.PdsLocations).Interface(), statem.Any.FullBytes,
-						"error with ASN.1 encoding, possibly a wrong ASN.1 string type was used"))
-			}
-			return etsiObj
+			return handleIdEtsiQcsQcEuPDS(statem)
 		} else if statem.Oid.Equal(IdEtsiQcsQcType) {
-			var qcType Etsi423QcType
-			qcType.isPresent = true
-			rest, err := asn1.Unmarshal(statem.Any.FullBytes, &qcType.TypeOids)
-			if len(rest) != 0 || err != nil {
-				return etsiBase{errorInfo: "error parsing IdEtsiQcsQcType extension statementInfo field", isPresent: true}
-			}
-			return qcType
+			return handleIdEtsiQcsQcType(statem)
+		} else if statem.Oid.Equal(IdEtsiPsd2Statem) {
+			return handleIdEtsiPsd2Statem(statem)
+		} else if statem.Oid.Equal(IdQcsPkixQCSyntaxV2) {
+			return handleIdQcsPkixQCSyntaxV2(statem)
 		} else {
 			return etsiBase{errorInfo: "", isPresent: true}
 		}
-
 	}
 
 	return etsiBase{errorInfo: "", isPresent: false}
+}
+
+func handleIdQcsPkixQCSyntaxV2(statem qcStatementWithInfoField) EtsiQcStmtIf {
+	var qcs2Statem DecodedQcS2
+	qcs2Statem.isPresent = true
+	if len(statem.Any.FullBytes) == 0 {
+		return qcs2Statem
+	}
+	rest, err := asn1.Unmarshal(statem.Any.FullBytes, &qcs2Statem.Decoded)
+	if err != nil {
+		AppendToStringSemicolonDelim(&qcs2Statem.errorInfo, "error parsing statement: "+err.Error())
+	}
+	if len(rest) != 0 {
+		AppendToStringSemicolonDelim(&qcs2Statem.errorInfo, "trailing bytes after QcStatement")
+	}
+	return qcs2Statem
+}
+
+func handleIdEtsiPsd2Statem(statem qcStatementWithInfoField) EtsiQcStmtIf {
+	var psd2Statem EtsiPsd2
+	psd2Statem.isPresent = true
+	rest, err := asn1.Unmarshal(statem.Any.FullBytes, &psd2Statem.DecodedPsd2Statm)
+	if len(rest) != 0 || err != nil {
+		return etsiBase{errorInfo: "error parsing IdEtsiPsd2Statem extension statementInfo field", isPresent: true}
+	}
+	if psd2Statem.DecodedPsd2Statm.CountryAndNCAId == "" || psd2Statem.DecodedPsd2Statm.NCAName == "" {
+		AppendToStringSemicolonDelim(&psd2Statem.errorInfo, "field has length 0")
+	}
+	for _, role := range psd2Statem.DecodedPsd2Statm.Roles {
+		if role.RoleOfPspName == "" {
+			AppendToStringSemicolonDelim(&psd2Statem.errorInfo, "field has length 0")
+		}
+	}
+	AppendToStringSemicolonDelim(&psd2Statem.errorInfo,
+		CheckAsn1Reencoding(reflect.ValueOf(psd2Statem.DecodedPsd2Statm).Interface(), statem.Any.FullBytes,
+			"error with ASN.1 encoding, possibly a wrong ASN.1 string type was used"))
+	return psd2Statem
+}
+
+func handleIdEtsiQcsQcType(statem qcStatementWithInfoField) EtsiQcStmtIf {
```

Comment:
Just noting: I'm ignoring lines 317-399; I'm expecting/assuming they're part of the refactor from 172-248. I'll save double-checking to the end, but if there's anything special to draw attention to, please let me know.